File: 2. Structural Design Patterns/Decorator/C#/DecoratorExample.cs
// Decorator Pattern Example in C#
// Notifier: Basic + Email + SMS + Slack decorators

using System;

namespace DecoratorExample
{
    interface INotifier { void Send(string message); }

    class BasicNotifier : INotifier
    {
        public void Send(string message)
        {
            Console.WriteLine($"Base notify: {message}");
        }
    }

    abstract class NotifierDecorator : INotifier
    {
        protected readonly INotifier wrappee;
        protected NotifierDecorator(INotifier wrappee) { this.wrappee = wrappee; }
        public virtual void Send(string message) { wrappee.Send(message); }
    }

    class EmailDecorator : NotifierDecorator
    {
        public EmailDecorator(INotifier w) : base(w) {}
        public override void Send(string message)
        {
            Console.WriteLine($"Email sent: {message}");
            base.Send(message);
        }
    }

    class SMSDecorator : NotifierDecorator
    {
        public SMSDecorator(INotifier w) : base(w) {}
        public override void Send(string message)
        {
            Console.WriteLine($"SMS sent: {message}");
            base.Send(message);
        }
    }

    class SlackDecorator : NotifierDecorator
    {
        public SlackDecorator(INotifier w) : base(w) {}
        public override void Send(string message)
        {
            Console.WriteLine($"Slack message sent: {message}");
            base.Send(message);
        }
    }

    class Program
    {
        static void Main()
        {
            INotifier notifier = new BasicNotifier();
            notifier = new EmailDecorator(notifier);
            notifier = new SMSDecorator(notifier);
            notifier = new SlackDecorator(notifier);
            notifier.Send("System maintenance at 2 AM");
        }
    }
}

File: 2. Structural Design Patterns/Decorator/Java/DecoratorExample.java
// Decorator Pattern Example in Java
// Notifier: Basic + Email + SMS + Slack decorators

interface Notifier { void send(String message); }

class BasicNotifier implements Notifier {
    public void send(String message){ System.out.println("Base notify: " + message); }
}

abstract class NotifierDecorator implements Notifier {
    protected final Notifier wrappee;
    protected NotifierDecorator(Notifier wrappee){ this.wrappee = wrappee; }
    public void send(String message){ wrappee.send(message); }
}

class EmailDecorator extends NotifierDecorator {
    public EmailDecorator(Notifier w){ super(w); }
    public void send(String message){ System.out.println("Email sent: " + message); super.send(message); }
}

class SMSDecorator extends NotifierDecorator {
    public SMSDecorator(Notifier w){ super(w); }
    public void send(String message){ System.out.println("SMS sent: " + message); super.send(message); }
}

class SlackDecorator extends NotifierDecorator {
    public SlackDecorator(Notifier w){ super(w); }
    public void send(String message){ System.out.println("Slack message sent: " + message); super.send(message); }
}

public class DecoratorExample {
    public static void main(String[] args){
        Notifier n = new BasicNotifier();
        n = new EmailDecorator(n);
        n = new SMSDecorator(n);
        n = new SlackDecorator(n);
        n.send("System maintenance at 2 AM");
    }
}
